openff/bespokefit/executor/services/coordinator/stages.py

```diff
@@ -290,44 +312,64 @@ def _regenerate_torsion_parameters(
         return new_smirks
 
     @staticmethod
-    async def _regenerate_parameters(
+    async def _generate_parameters(
         fragmentation_result: FragmentationResult,
         input_schema: BespokeOptimizationSchema,
     ):
         """
-        Regenerate any place holder torsion parameters in the input schema and add them to the force field while
-        removing old values. This edits the input schema in place.
+        Generate a list of parameters which are to be optimised, these are added to the input force field.
+        The parameters are also added to the parameter list in each stage corresponding to the stage where they will be fit.
         """
 
         initial_force_field = ForceFieldEditor(input_schema.initial_force_field)
+        new_parameters = []
 
-        torsion_parameters = OptimizationStage._regenerate_torsion_parameters(
-            fragmentation_result=fragmentation_result,
-            initial_force_field=input_schema.initial_force_field,
-        )
+        target_smirks = {*input_schema.target_smirks}
+        if SMIRKSType.ProperTorsions in target_smirks:
+            target_smirks.remove(SMIRKSType.ProperTorsions)
 
-        torsion_handler = initial_force_field.force_field["ProperTorsions"]
+            torsion_parameters = OptimizationStage._generate_torsion_parameters(
+                fragmentation_result=fragmentation_result,
+                input_schema=input_schema,
+            )
+            new_parameters.extend(torsion_parameters)
+
+        if len(target_smirks) > 0:
+            smirks_gen = SMIRKSGenerator(
+                initial_force_field=input_schema.initial_force_field,
+                generate_bespoke_terms=input_schema.generate_bespoke_terms,
+                target_smirks=[*target_smirks],
+                smirks_layers=1,
+            )
 
-        new_parameters = []
-        # now we can remove all of the place holder terms
-        for old_parameter in input_schema.stages[0].parameters:
-            if isinstance(old_parameter, ProperTorsionSMIRKS):
-                # remove from the old force field
-                del torsion_handler.parameters[old_parameter.smirks]
-            else:
-                # we want to keep the other parameters
-                new_parameters.append(old_parameter)
-
-        # now add our new parameters to the force field and the schema list
-        for torsion_parameter in torsion_parameters:
-            new_parameters.append(
-                ProperTorsionSMIRKS(
-                    smirks=torsion_parameter.smirks, attributes={"k1", "k2", "k3", "k4"}
-                )
+            parameters = smirks_gen.generate_smirks_from_molecule(
+                molecule=input_schema.molecule
```

Comment:
Should this logic be based on the QC data that's available rather than fragments? This way it should capture parameters for the new fragments as well as for any user provided QC data